src/pybind/mgr/cephadm/services/cephadmservice.py

```diff
@@ -14,6 +14,48 @@
 
 logger = logging.getLogger(__name__)
 
+ServiceSpecs = TypeVar('ServiceSpecs', bound=ServiceSpec)
+
+
+class CephadmDaemonSpec(Generic[ServiceSpecs]):
+    # typing.NamedTuple + Generic is broken in py36
+    def __init__(self, host, daemon_id,
+                 spec: Optional[ServiceSpecs]=None,
+                 network: Optional[str]=None,
+                 keyring: Optional[str]=None,
+                 extra_args: Optional[List[str]]=None,
+                 extra_config: Optional[Dict[str, Any]]=None,
+                 daemon_type: Optional[str]=None):
+        """
+        Used for
+        * deploying new daemons. then everything is set
+        * redeploying existing daemons, then only the first three attrs are set.
+
+        Would be great to have a consistent usage where all properties are set.
+        """
+        self.host = host
+        self.daemon_id = daemon_id
+        daemon_type = daemon_type or (spec.service_type if spec else None)
+        assert daemon_type is not None
+        self.daemon_type: str = daemon_type
+
+        # would be great to have the spec always available:
+        self.spec: Optional[ServiceSpecs] = spec
+
+        # mons
+        self.network = network
+
+        # for run_cephadm.
+        self.keyring: Optional[str] = keyring
+
+        # For run_cephadm. Would be great to have more expressive names.
+        self.extra_args: List[str] = extra_args or []
+        self.extra_config: Dict[str, Any] = extra_config or {}
+
+
+    def name(self) -> str:
+        return '%s.%s' % (self.daemon_type, self.daemon_id)
+
 
 class CephadmService(metaclass=ABCMeta):
```

Comment:
CephadmService class ... This abstraction seems not to be for a service, it seems to be a "factory" class for create daemons.... In my opinion the functionality included here should be moved to  a "CephadmDaemon" class

The CephadmService class should model a "Service"... and in a Service what we have is:
- one or more daemons providing the service
- health of the service
- tools to scale up/down the service
- tools to provide a virtual ip ( for load balancing and HA)

Reply:
hm that's not totally correct. Like e.g. `get_active_daemon` which returns the active daemon for a set of daemons belonging to a service. But sure, some methods work on daemons belonging to a service.